File: day2/part1/solution.exs
defmodule Day2Part1 do
  def solve(input_str) do
    ranges = parse_ranges(input_str)
    max_id = find_max_id(ranges)
    
    1..max_id
    |> Stream.map(&to_string/1)
    |> Stream.map(&(&1 <> &1))
    |> Stream.map(&String.to_integer/1)
    |> Stream.take_while(&(&1 <= max_id))
    |> Enum.reduce(MapSet.new(), fn invalid_id, acc ->
      if in_any_range?(invalid_id, ranges) do
        MapSet.put(acc, invalid_id)
      else
        acc
      end
    end)
    |> Enum.sum()
  end

  defp parse_ranges(input_str) do
    input_str
    |> String.trim()
    |> String.split(",")
    |> Enum.map(fn r_str ->
      [start_str, end_str] = String.split(r_str, "-")
      {String.to_integer(start_str), String.to_integer(end_str)}
    end)
  end

  defp find_max_id(ranges) do
    ranges
    |> Enum.map(&elem(&1, 1))
    |> Enum.max()
  end

  defp in_any_range?(id, ranges) do
    Enum.any?(ranges, fn {start, finish} ->
      id >= start and id <= finish
    end)
  end
end

input = File.read!("day2/part1/input.txt")
result = Day2Part1.solve(input)
IO.puts("The sum of all invalid IDs is: #{result}")

File: day2/part1/solution.rs
use std::collections::HashSet;
use std::fs;
use std::io;

fn parse_ranges(input_str: &str) -> (Vec<(u64, u64)>, u64) {
    let mut ranges = Vec::new();
    let mut max_id = 0;
    for r_str in input_str.trim().split(',') {
        if r_str.is_empty() {
            continue;
        }
        let parts: Vec<&str> = r_str.split('-').collect();
        let start = parts[0].parse().unwrap();
        let end = parts[1].parse().unwrap();
        ranges.push((start, end));
        if end > max_id {
            max_id = end;
        }
    }
    (ranges, max_id)
}

fn generate_invalid_ids(max_id: u64) -> Vec<u64> {
    let mut invalid_ids = Vec::new();
    let mut base: u64 = 1;
    loop {
        let s_base = base.to_string();
        let s_invalid = s_base.clone() + &s_base;
        let n_invalid: u64 = match s_invalid.parse() {
            Ok(n) => n,
            Err(_) => break, // Should not happen with this logic
        };

        if n_invalid > max_id {
            if s_base.len() > (base -1).to_string().len() && base > 1 {
                 break;
            }
        }
        
        invalid_ids.push(n_invalid);
        base += 1;
    }
    invalid_ids
}

fn solve(input_str: &str) -> u64 {
    let (ranges, max_id) = parse_ranges(input_str);
    let potential_ids = generate_invalid_ids(max_id);

    let mut found_invalid_ids = HashSet::new();

    for &invalid_id in &potential_ids {
        for &(start, end) in &ranges {
            if invalid_id >= start && invalid_id <= end {
                found_invalid_ids.insert(invalid_id);
                break; 
            }
        }
    }

    found_invalid_ids.iter().sum()
}

fn main() -> io::Result<()> {
    let input_str = fs::read_to_string("day2/part1/input.txt")?;
    let result = solve(&input_str);
    println!("The sum of all invalid IDs is: {}", result);
    Ok(())
}